### ISSUE TITLE
Issue getting simultaneous Deepdream working alongside Neural-Style

### ISSUE THREAD

**Issue by ProGamerGov** [jcjohnson/neural-style#389]
I am trying to get [Fast-Neural-Style's DeepDreamLoss feature](https://github.com/jcjohnson/fast-neural-style/blob/master/fast_neural_style/DeepDreamLoss.lua) working in Neural-Style.

I currently have this modified version of Neural-Style setup: https://gist.github.com/ProGamerGov/9e7b109458828a1f9e7cf995a4d2fd4b, but the Deepdream loss remains 0 no matter what I do. Here is the command line output when using the Deepdream related parameters: https://gist.github.com/ProGamerGov/fa20f22cdd5a89c2ef998b2f6b716dd1

I tried to replicate the style/content loss functions with the Deepdream loss function as well as remaining faithful to the Fast-Neural-Style code. But I am not sure where exactly the Deepdream code should connect with the loss functions as the Fast-Neural-Style code is setup differently than Neural-Style's code.

**Comment by jcjohnson** [comment]
@ProGamerGov In `DeepDreamLoss:updateOutput` you are not updating self.loss, so it will always be 0. Computing the loss doesn't actually matter for creating images; it's just a convenient way for you to track progress. What actually matters is that you compute the gradient with respect to the loss, which your code seems to do correctly.

Therefore even though the loss remains 0, I think that your code should be able to produce DeepDream images if you get the hyperparameters right.

**Comment by ProGamerGov** [comment]
@jcjohnson Thanks for the reply. It seems to work now that I fixed the [name of the deepdream_weights parameter](https://gist.github.com/ProGamerGov/9e7b109458828a1f9e7cf995a4d2fd4b#file-neural_style_dd-lua-L155) on line 155.  

![](https://i.imgur.com/HVfwqfp.png)

Though as seen in the image above, and [this image](https://i.imgur.com/IDcNVVN.png), there seems to be a lot of blue and green lines. I think this may be related to the `max_grad` input, but I am not sure what value it is supposed to be. I am currently assigning the the `normalize_gradients` value to the `max_grad` input. 

Also, is implementing the other Deepdream parameters like `-num_octaves` and `-octave_scale` possible?

**Comment by ProGamerGov** [comment]
@jcjohnson Thanks for the reply. It turns out that I misspelled the `-deepdream_weights` parameter I was passing to `nn.DeepDreamLoss `, and that's why the code was not working. After some experimentation, I believe it now works like slow_neural_style.lua's DeepDream, but with all the features of the regular neural_style.lua. 

I cleaned up the script by removing the unneeded parts that I originally added for DeepDream: https://gist.github.com/ProGamerGov/ff2715386d738e8245335b1a95617a1b

> In `DeepDreamLoss:updateOutput` you are not updating self.loss, so it will always be 0. Computing the loss doesn't actually matter for creating images; it's just a convenient way for you to track progress.  

The style and content loss functions update self.loss with the criterion. Is it possible to update the Deepdream self.loss without the criterion, and is the information obtained from doing so as useful as it is for the content and style loss functions?

**Comment by ProGamerGov** [comment]
I am attempting to get octaves working, but I `image.scale` does not appear to work inside `DeepDreamLoss:updateGradInput` using the code below. 

```
local octaves = {}
     octaves[self.octave_n] = self.gradInput:add(self.clipped, -self.max_grad):div(self.max_grad) 
     for i=self.octave_n-1,1,-1 do
        local _,h,w = unpack(self.base_img:size():totable())
        octaves[i] = image.scale(octaves[i+1], math.ceil((1/self.octave_scale)*w), math.ceil((1/self.octave_scale)*h),'simple')
     end
```


From `DeepDreamLoss:__init`:

```
  self.octave_n = octave_n or 4
  self.octave_scale = octave_scale or 1.4

  self.base_img = base_img
```

base_img is the resized input image: 

```
-- DeepDream contrast fix 
  if params.deepdream_contrast ~= '' then
    for i = 1, #deepdream_losses do
      deepdream_losses[i].mode = 'contrast'

      local content_image_size = image.load(params.content_image, 3)
      content_image_size = image.scale(content_image_size, params.image_size, 'bilinear')
      deepdream_losses[i].base_img = content_image_size
    end
  end

```

The full modified neural_style.lua can be found here: https://gist.github.com/ProGamerGov/ad143adf1b7d4eb2bd2d177153abc00e

**Comment by htoyryla** [comment]
@ProGamerGov 
at this line you effectively make gradInput a copy of gradOutput, which is probably why you avoid the errors, but you also lose everything you have done with gradInput so far.
